### ISSUE TITLE
completion working in ignored scopes

### ISSUE THREAD

**Issue by AmjadHD** [tomv564/LSP#615]
completion is working in string and comments

* OS: Windows 10
* How you installed LSP: git
* Minimal reproduction steps
1. open a python file
2. type `"str."`

**Comment by tomv564** [comment]
The logic is still there, but it's bypassed if there is a prefix:
```python
if prefix != "" and self.view.match_selector(locations[0], NO_COMPLETION_SCOPES):
     # return empty results
```

I think we'll have to dig through the history to know why prefixes bypass the filter, but I believe there was a request for completions to work within strings (eg template literals)

**Comment by frou** [comment]
An even more obviously wrong situation is typing a comment and putting a period at the end of the sentence, and having code completion pop up. This is happening with Python and `pyls`.

**Comment by tomv564** [comment]
Related request was #301, I think we can fix this with a more advanced scope selector that disallows strings but recognises some common scopes used on placeholders/variables in template strings. Any help finding what scopes these are for other languages (I documented for typescript in the related issue) would be helpful!

Setting`only_show_lsp_completions=True` currently also will disable the sublime word completion within comments and strings. This feels a bit incorrect to me and LSP shouldn't care about what happens within scopes for which it doesn't want to handle completion requests. What do you think?

**Comment by digulla** [comment]
I think languages which support string templates shouldn't use the `string` but a `template` scope instead. `string` should only be used for (dump) text literals which allow at most escape sequences. So

    "First Line|

Pressing Ctrl+Space here should give me word completion with the prefix `Line` plus a list of escape sequences (\n, \r, \t, \\, single backslash to continue the line), etc. In Python and C, this might include helpers like `%s`, `%d`.

On the other hand, in Python

    f"The value is {|

should give me all local variables since this is now a template and I'm inside of an expression. Everything outside of `{}` should be treated like dump text (like above).

**Comment by tomv564** [comment]
I decided to follow sublime's defaults in the recently merged pull request #723:
* Completions in non-block strings
* No completions in comments

You can change sublime's `auto_complete_selector` globally or per syntax to eg. turn off completions in strings.

I realise there is perhaps a wish for sublime word completion to work in scopes where LSP completion is disabled - let me hear your argument for this!
